#-------------------------------------------------------------------------------
# Author: David Prince
# Project: NER0008751 (Obj3) 
# Analysis: mRNA-seq
# Subsection: Differential expression analysis.
# Tasks: Produce a summary of each differential expression analysis consisting 
# of a MA plot and a summary table.
#-------------------------------------------------------------------------------
# Inputs:
# Dmel_6_22_97_transcripts2genes.txt file and Kallisto abundances.

# Outputs:
# PDF figure of MA plot and summary table of DEGs at different log-fold 
# changes (LFCs).
#-------------------------------------------------------------------------------

# LOADING PACKAGES ----

library(tidyr)  # replace_na().
library(ggplot2)  # ggplot(), ggsave().
library(ggpubr)  # ggtexttable(), ggarrange(), annotate_figure().
# DESeq2 and tximport loaded via the scripts in the 
# "Custom functions" section.

# LOAD DATA ----
# NOTE: This script assumes that the working directory is the 01_scripts 
# subdirectory.

# Transcript to Gene File ----

setwd("../00_data/10_transcripts2genes")

t2g <- read.table(file = "Dmel_6_22_97_transcripts2genes.txt",
                  header = FALSE,
                  col.names = c("TXNAME",
                                "GENEID"))

# LOAD CUSTOM FUNCTIONS ----

setwd("../../01_scripts")

source("f_BuildDDSForDEAnalysis.R")

# FUNCTION DEFINITIONS ----

GenerateDEGSummary <- function (x, y) {
  # Generates summary statistics for the number of statistically significant
  # differentially expressed genes (DEGs) for a given tissue, treatment and 
  # log-fold change (LFC).
  #
  # Args:
  #   x: string denoting which tissue is to be analysed ("fatbody", "head"  
  #      or "ovaries").
  #   y: string denoting the name of treatment to generate a summary for 
  #       ("M" or "H").
  #
  # Returns:
  #   Character vector summarising the results for a given tissue and treatment.
  
  # Set variables.
  
  if (x == "head") {
    dds <- ddsHead
  } else if (x == "fatbody") {
    dds <- ddsFatBody
  } else if (x == "ovaries") {
    dds <- ddsOvaries
  } else {
    stop('Argument x must be "fatbody", "head" or "ovaries".')
  }
  
  if (y == "M") {
    a <- "M_RNA2"
    b <- "M_RNA1"
    treatment <- "medium"
  } else if (y == "H") {
    a <- "H_RNA2"
    b <- "H_RNA1"
    treatment <- "high"
  } else {
    stop('Argument y must equal "M" or "H".')
  }
  
  # Extract results for LFC > 0.
  
  extractedResultsLFC0 <- results(dds, contrast = c("condition", a, b),
                              alpha = 0.05, lfcThreshold = 0)
  
  # Extract results for LFC > 1.
  
  extractedResultsLFC1 <- results(dds, contrast = c("condition", a, b),
                                  alpha = 0.05, lfcThreshold = 1)
  
  # Generate summary object.
  
  resultsSummary <- capture.output(c(summary(extractedResultsLFC0)),
                                   summary(extractedResultsLFC1))
  
  # Format resultsSummary for improved readability.
  
  resultsSummary <- resultsSummary[c(2:5, 16:17)]
  
  # Generate description of results extracted.
  
  resultsDescription <- 
    paste0("Results for: tissue = ", x, ", treatment = ", treatment, ".")
  
  # Combine description and results.
  
  resultsToReturn <- c(resultsDescription, resultsSummary)
  
  # Return summary.
  
  return(resultsToReturn)
  
}

MakeMAPlot <- function(x, y) {
  # Generates a MA plot for DEGs from a given tissue/treatment.
  #
  # Args:
  #   x: string denoting which tissue is to be analysed ("fatbody", "head"  
  #      or "ovaries").
  #   y: string denoting the name of treatment to generate a summary for 
  #       ("M" or "H").
  #
  # Returns:
  #   A MA plot generated by ggplot2.
  
  # Assign variables based on x.
  
  if (x == "head") {
    dds <- ddsHead
  } else if (x == "fatbody") {
    dds <- ddsFatBody
  } else if (x == "ovaries") {
    dds <- ddsOvaries
  } else {
    stop('Argument x must be "fatbody", "head" or "ovaries".')
  }
  
  if (y == "M") {
    condition <- c("condition", "M_RNA2", "M_RNA1")
  } else if (y == "H") {
    condition <- c("condition", "H_RNA2", "H_RNA1")
  } else {
    stop('Argument y must equal "M" or "H".')
  }
  
  # LFC shrink.
  
  resLFCShrink <- lfcShrink(dds, contrast = condition, type = "ashr")
  
  # Make data.frame.
  
  res_df <- as.data.frame(resLFCShrink)
  
  # Add significance column.
  
  res_df <- dplyr::mutate(res_df, significant = res_df$padj < 0.05)
  
  # Convert significance column to character rather than logical.
  
  res_df$significant <- as.character(res_df$significant)
  
  # Change NAs in significance column to "FALSE".
  
  res_df$significant <- replace_na(res_df$significant, "FALSE")
  
  # Produce MA plot.
  
  maPlot <- ggplot(res_df, aes(x = log2(baseMean), y = log2FoldChange, 
                               colour = significant)) +
    geom_point(size = 0.85, shape = 19) +  # Controls the point size and shape.
    scale_colour_brewer(palette = "Set1") +  # Controls colours.
    ylim(-4, 4) +  # Controls size of y-axis.
    theme_bw()
  
  # Return MA plot.
  
  return(maPlot)
  
}

MakeFigure <- function(x) {
  # Produces a figure consisting of two MA plots and two differential expression
  # summaries for the stated tissue.
  #
  # Args:
  #   x: string denoting which tissue is to be analysed ("fatbody", "head"  
  #      or "ovaries").
  #
  # Returns:
  #   A ggplot figure arranged by ggpubr consisting of four subplots; two MA plots
  #   and two tables summarising numbers of DEGs (one for each of the treatments).
  
  # Assign variables based on argument x.
  
  if (x == "head") {
    maM <- headMMA
    maH <- headHMA
    sumM <- ggtexttable(headMSum)
    sumH <- ggtexttable(headHSum)
  } else if (x == "fatbody") {
    maM <- fatBodyMMA
    maH <- fatBodyHMA
    sumM <- ggtexttable(fatBodyMSum)
    sumH <- ggtexttable(fatBodyHSum)
  } else if (x == "ovaries") {
    maM <- ovariesMMA
    maH <- ovariesHMA
    sumM <- ggtexttable(ovariesMSum)
    sumH <- ggtexttable(ovariesHSum)
  } else {
    stop('Argument x must be "fatbody", "head" or "ovaries".')
  }
  
  #  Combine text and graph.
  
  newPlot <- ggarrange(maM, sumM, maH, sumH, 
                       labels = c("A", "B", "C", "D"), 
                       ncol = 2, nrow = 2)
  
  figure <- annotate_figure(newPlot,
                            top = x)
  
  # Return figure.
  
  return(figure)
  
}

# EXECUTED STATEMENTS ----

# Build dds Objects for Each Tissue ----

# Head.

ddsHead <- BuildDDSForDEAnalysis("head")

# Fat body.

ddsFatBody <- BuildDDSForDEAnalysis("fatbody")

# Ovaries.

ddsOvaries <- BuildDDSForDEAnalysis("ovaries")

# Generate Summaries of Differential Expression Analysis ----

# Head.

headMSum <- GenerateDEGSummary("head", "M")

headHSum <- GenerateDEGSummary("head", "H")

# Fat body.

fatBodyMSum <- GenerateDEGSummary("fatbody", "M")

fatBodyHSum <- GenerateDEGSummary("fatbody", "H")

# Ovaries.

ovariesMSum <- GenerateDEGSummary("ovaries", "M")

ovariesHSum <- GenerateDEGSummary("ovaries", "H")

# Generate MA Plots ----

# Head.

headMMA <- MakeMAPlot("head", "M")

headHMA <- MakeMAPlot("head", "H")

# Fat body.

fatBodyMMA <- MakeMAPlot("fatbody", "M")

fatBodyHMA <- MakeMAPlot("fatbody", "H")

# Ovaries.

ovariesMMA <- MakeMAPlot("ovaries", "M")

ovariesHMA <- MakeMAPlot("ovaries", "H")

# Make Figures ----

# Head.

headFigure <- MakeFigure("head")

# Fat body.

fatBodyFigure <- MakeFigure("fatbody")

# Ovary.

ovariesFigure <- MakeFigure("ovaries")

# Save Figures ----

# Head.

# Create directory.

dir.create("../02_outputs/02_head/30_DESeq2_summary_figure")
# Will produce a warning if directory already exists.

# Change directory

setwd("../02_outputs/02_head/30_DESeq2_summary_figure")

# Save figure.

ggsave("00_NER0008751_obj3_exp3_head_deseq2_summary_figure.pdf",
       headFigure, width = 29.7, height = 21, units = "cm")

# Fat body.

# Create directory.

dir.create("../../03_fatbody/30_DESeq2_summary_figure")
# Will produce a warning if directory already exists.

# Change directory.

setwd("../../03_fatbody/30_DESeq2_summary_figure")

# Save figure.

ggsave("00_NER0008751_obj3_exp3_fatbody_deseq2_summary_figure.pdf",
       fatBodyFigure, width = 29.7, height = 21, units = "cm")

# Ovaries.

# Create directory.

dir.create("../../01_ovaries/30_DESeq2_summary_figure")
# Will produce a warning if directory already exists.

# Change directory.

setwd("../../01_ovaries/30_DESeq2_summary_figure")

# Save figure.

ggsave("00_NER0008751_obj3_exp3_ovaries_deseq2_summary_figure.pdf",
       ovariesFigure, width = 29.7, height = 21, units = "cm")
